### PR TITLE
Bug 1162843 - Allow searching for multiple non-consecutive words in a title

### PR DESCRIPTION
**Pull request by Johennes**
I'd appreciate some feedback here. I have a feeling my code might be a little too compact and could be hard to read. I could make it more expressive by using a dedicated `for` loop. I'm also thinking about adding unit tests for this but I wanted to get some early opinion first.

### CODE REVIEW

**Comment by rnewman** [comment]
I'd want to see a query plan and some perf measurements for this before moving forward. Best-case this is still a single table scan, but it's doing a linear string match for each substring, and who knows what the worst case is.

Bug 1194325 is likely to make our perf slightly worse, not better, so we have to be conscious of the costs and benefits.

I'd also like someone to figure out what our FTS story is.

It's also worth thinking about whether or how this can interact with ranking, and whether it's worthwhile introducing the concept of matching the whole entered substring.

**Comment by rnewman** [comment]
To flesh that out a little more: whole-entered-substring is what matters when searching in your history for the band "The The". (This particular example is also about noise words, but you get the gist.)

File: Storage/SQL/SQLiteHistory.swift [hunk 1]
```diff
@@ -323,10 +323,12 @@ extension SQLiteHistory: BrowserHistory {
         let whereClause: String
         let whereFragment = (whereData == nil) ? "" : " AND (\(whereData!))"
         if let filter = filter {
-            args = ["%\(filter)%", "%\(filter)%"]
-
+            // Split filter on whitespace and set up where clause for every word.
+            let words = filter.componentsSeparatedByCharactersInSet(NSCharacterSet.whitespaceCharacterSet()).filter { !$0.isEmpty }
```
Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: You probably want to remove duplicates here, and perhaps even substrings.

File: Storage/SQL/SQLiteHistory.swift [hunk 2]
```diff
@@ -323,10 +323,12 @@ extension SQLiteHistory: BrowserHistory {
...
             // No deleted item has a URL, so there is no need to explicitly add that here.
-            whereClause = " WHERE ((\(TableHistory).url LIKE ?) OR (\(TableHistory).title LIKE ?)) \(whereFragment)"
+            let filterFragment = "(" + " OR ".join(words.map { _ in "(\(TableHistory).url LIKE ?) OR (\(TableHistory).title LIKE ?)" }) + ")"
```
Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: I think this should be joining with `AND`, right? You want to match every entered word in either the URL or the title, not any of them.

Comment (Johennes) [inline Storage/SQL/SQLiteHistory.swift]: You're right. Oh my, what a stupid mistake, sorry.

**Comment by Johennes** [comment]
So the idea would be to match either the whole filter string or a combination of it's individual words minus duplicates and substrings? I will try to get that in and then take a shot at getting the SQL query plan and compiling some performance data.

**Comment by Johennes** [comment]
I got the suggested changes in. The algorithm for dropping duplicates and substrings is not very sophisticated but I think it should do for a basic test given that the concerns were about SQL (not Swift) performance.

**Comment by Johennes** [comment]
I finally got around to investigating the performance. This is an example of the resulting select statement for the search string `"firefox is my favorite browser"` without splitting (as in `master`)

``` sql
SELECT historyID, url, title, guid, domain_id, domain, localVisitDate, remoteVisitDate, localVisitCount, remoteVisitCount, iconID, iconURL, iconDate, iconType, iconWidth
FROM (
  SELECT historyID, url, title, guid, domain_id, domain, max(localVisitDate) AS localVisitDate, max(remoteVisitDate) AS remoteVisitDate, sum(localVisitCount) AS localVisitCount, sum(remoteVisitCount) AS remoteVisitCount
  FROM (
    SELECT history.id AS historyID, history.url AS url, title, guid, domain_id, domain, COALESCE(max(case visits.is_local when 1 then visits.date else 0 end), 0) AS localVisitDate, COALESCE(max(case visits.is_local when 0 then visits.date else 0 end), 0) AS remoteVisitDate, COALESCE(sum(visits.is_local), 0) AS localVisitCount, COALESCE(sum(case visits.is_local when 1 then 0 else 1 end), 0) AS remoteVisitCount
	FROM history INNER JOIN domains ON domains.id = history.domain_id INNER JOIN visits ON visits.siteID = history.id 
	WHERE ((history.url LIKE "%firefox is my favorite browser%") OR (history.title LIKE "%firefox is my favorite browser%"))
	GROUP BY historyID)
  WHERE ((localVisitCount + remoteVisitCount) > 0)
  GROUP BY historyID
  ORDER BY (localVisitCount * (5 + localVisitCount)) * max(1, 100 * 225 / (((1440319682374770 - (localVisitDate)) / 86400000000.0) * ((1440319682374770 - (localVisitDate)) / 86400000000.0) + 225)) + remoteVisitCount * max(1, 100 * 225 / (((1440319682374794 - (remoteVisitDate)) / 86400000000.0) * ((1440319682374794 - (remoteVisitDate)) / 86400000000.0) + 225)) DESC  LIMIT 100 )
LEFT OUTER JOIN view_history_id_favicon ON historyID = view_history_id_favicon.id
```

and with splitting (as in this pull request)

``` sql
SELECT historyID, url, title, guid, domain_id, domain, localVisitDate, remoteVisitDate, localVisitCount, remoteVisitCount, iconID, iconURL, iconDate, iconType, iconWidth 
FROM (
  SELECT historyID, url, title, guid, domain_id, domain, max(localVisitDate) AS localVisitDate, max(remoteVisitDate) AS remoteVisitDate, sum(localVisitCount) AS localVisitCount, sum(remoteVisitCount) AS remoteVisitCount
  FROM (
    SELECT history.id AS historyID, history.url AS url, title, guid, domain_id, domain, COALESCE(max(case visits.is_local when 1 then visits.date else 0 end), 0) AS localVisitDate, COALESCE(max(case visits.is_local when 0 then visits.date else 0 end), 0) AS remoteVisitDate, COALESCE(sum(visits.is_local), 0) AS localVisitCount, COALESCE(sum(case visits.is_local when 1 then 0 else 1 end), 0) AS remoteVisitCount
	FROM history INNER JOIN domains ON domains.id = history.domain_id INNER JOIN visits ON visits.siteID = history.id 
	WHERE (((history.url LIKE "%firefox is my favorite browser%") OR (history.title LIKE "%firefox is my favorite browser%"))
	      OR ((history.url LIKE "%firefox%") OR (history.title LIKE "%firefox%"))
		  AND ((history.url LIKE "%is%") OR (history.title LIKE "%is%"))
		  AND ((history.url LIKE "%my%") OR (history.title LIKE "%my%"))
		  AND ((history.url LIKE "%favorite%") OR (history.title LIKE "%favorite%"))
		  AND ((history.url LIKE "%browser%") OR (history.title LIKE "%browser%")))
	GROUP BY historyID)
  WHERE ((localVisitCount + remoteVisitCount) > 0)
  GROUP BY historyID
  ORDER BY (localVisitCount * (5 + localVisitCount)) * max(1, 100 * 225 / (((1440319482893215 - (localVisitDate)) / 86400000000.0) * ((1440319482893215 - (localVisitDate)) / 86400000000.0) + 225)) + remoteVisitCount * max(1, 100 * 225 / (((1440319482893270 - (remoteVisitDate)) / 86400000000.0) * ((1440319482893270 - (remoteVisitDate)) / 86400000000.0) + 225)) DESC  LIMIT 100 )
LEFT OUTER JOIN view_history_id_favicon ON historyID = view_history_id_favicon.id
```

The SQLite query plan for both is identical and looks like this

```
2|0|0|SCAN TABLE history
2|1|1|SEARCH TABLE domains USING INTEGER PRIMARY KEY (rowid=?)
2|2|2|SEARCH TABLE visits USING COVERING INDEX idx_visits_siteID_is_local_date (siteID=?)
1|0|0|SCAN SUBQUERY 2
1|0|0|USE TEMP B-TREE FOR GROUP BY
1|0|0|USE TEMP B-TREE FOR ORDER BY
4|0|0|SCAN TABLE favicon_sites USING COVERING INDEX sqlite_autoindex_favicon_sites_1
4|1|1|SEARCH TABLE favicons USING INTEGER PRIMARY KEY (rowid=?)
3|0|0|SCAN TABLE history USING COVERING INDEX sqlite_autoindex_history_2
3|1|1|SEARCH SUBQUERY 4 USING AUTOMATIC COVERING INDEX (siteID=?)
0|0|0|SCAN SUBQUERY 1
0|1|1|SEARCH SUBQUERY 3 USING AUTOMATIC COVERING INDEX (id=?)
```

The select that is matching against the search string has the ID 2 and appears at the very top. It results in a single full table scan of the history table.

I also measured the query time using the command line interface of SQLite. The unsplit version averages around 17ms whereas the split version takes about 21ms. This is of course a unique example but it, nevertheless, gives some hints about the impact we'd have to expect.

**Comment by rnewman** [comment]
We reworked the frecency calculation recently, so I decided to try this out using the current query. Timings against my own history DB are comments in the SQL. My history has about 8000 rows, with about 700 matches.

```
SELECT *, (((2 + localVisitCount) * (2 + localVisitCount)) * max(2, 100 * 225 / (((1439929399333085 - localVisitDate) / 86400000000.0) * ((1439929399333085 - localVisitDate) / 86400000000.0) + 225)) + remoteVisitCount * max(1, 100 * 110 / (((1439929399333104 - remoteVisitDate) / 86400000000.0) * ((1439929399333104 - remoteVisitDate) / 86400000000.0) + 110))) AS frecency
FROM (
SELECT history.id AS historyID, history.url AS url, title, guid,
domain_id, domain,
COALESCE(max(case visits.is_local when 1 then visits.date else 0 end), 0) AS localVisitDate,
COALESCE(max(case visits.is_local when 0 then visits.date else 0 end), 0) AS remoteVisitDate,
COALESCE(sum(visits.is_local), 0) AS localVisitCount,
COALESCE(sum(case visits.is_local when 1 then 0 else 1 end), 0) AS remoteVisitCount
FROM history
INNER JOIN domains ON domains.id = history.domain_id
INNER JOIN visits ON visits.siteID = history.id 
WHERE
(history.is_deleted = 0)
AND
-- Run Time: real 0.025 user 0.015640 sys 0.003440
((history.url LIKE "%Google%") OR (history.title LIKE "%Google%")) AND ((history.url LIKE "%Search%") OR (history.title LIKE "%Search%"))

-- Run Time: real 0.030 user 0.016024 sys 0.004565
--((history.url LIKE "%Google Search%") OR (history.title LIKE "%Google Search%"))
GROUP BY historyID)
WHERE (localVisitCount > 0) OR (remoteVisitCount > 0)
ORDER BY frecency DESC LIMIT 1000;
```

You'll see that the new "and-or" approach actually takes *less* time. I can think of hypotheses for that, but none that make much sense!

**Comment by rnewman** [comment]
A pathological example, searching for "a www":

```
Split (2,900 results):
Run Time: real 0.073 user 0.028277 sys 0.003515

Combined (old) (0 results):
Run Time: real 0.007 user 0.007046 sys 0.000705
```

So you can see that the split approach is much slower if one of the terms — probably the first — is very non-specific, and the combined approach naturally is quicker if it doesn't match many rows.

**Comment by Johennes** [comment]
Concerning the pathological case, I was about to propose to order the split filter words by their length but it actually seems that this doesn't make much of a difference. Comparing the average of 20 runs of

``` sql
WHERE (((history.url LIKE "%a www%") OR (history.title LIKE "%a www%")) OR
       ((history.url LIKE "%a%") OR (history.title LIKE "%a%")) AND
       ((history.url LIKE "%www%") OR (history.title LIKE "%www%")))
```

against

``` sql
WHERE (((history.url LIKE "%a www%") OR (history.title LIKE "%a www%")) OR
       ((history.url LIKE "%www%") OR (history.title LIKE "%www%")) AND
       ((history.url LIKE "%a%") OR (history.title LIKE "%a%")))
```

results in a query time difference of merely about 1ms (total query time about 55ms) for me. This also stays true when I use a more specific second word like in "a sophisticated".

Would it be an option to omit short words (e.g. <= 3 characters) from the split part of the where clause? In this case, "a www" would result in just

``` sql
WHERE (((history.url LIKE "%a www%") OR (history.title LIKE "%a www%")))
```

which is the current unsplit form.

I totally agree about the tests. I just wanted to nail down the logic (and whether or not to move forward) first before investing into them.

**Comment by Johennes** [comment]
I'm gonna rebase this on master and try to get some basic tests going within the next days.

**Comment by rnewman** [comment]
It's probably not worth introducing additional complexity to the splitter. Most of the query time is proportional to the number of results, because once we're doing a table scan — and we are — the expense is in ordering and preparing the cursor.

**Comment by Johennes** [comment]
I've been thinking that by removing short/noise words (which makes the query more specific) we would also reduce the number of results but you're probably right - especially since I'm not sure whether this would work out in different languages.

I'll have to break my promise of updating the PR _within the next days_. I'll probably do it on the weekend, sorry.

**Comment by Johennes** [comment]
I finally rebased the commit on master and migrated it to the latest Swift syntax. I also did some refactoring and added unit tests. I was unable to actually test the _Client_ scheme, however. There are syntax errors in some test files and even after fixing them, tests would always crash with an unexpected nil during unwrapping. I verified that my tests succeed with a separate sample project.

**Comment by Johennes** [comment]
Any further thoughts on this? Anything I should change or improve?

**Comment by rnewman** [comment]
I finally — finally! — cleaned this up, fixed the behavior, and landed this. Thanks for your patience, Johennes!

5238e87

**Comment by Johennes** [comment]
Awesome! :)

**Pull request closed without merging**